data_structures/src/chain.rs

```diff
@@ -784,6 +882,236 @@ pub enum InventoryItem {
+                tally,
+            })
+        } else {
+            Err("Cannot persist unfinished data request (with no Tally)")
+        }
+    }
+}
+
+/// List of outputs related to a data request
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct DataRequestInfo {
+    /// List of commitments to resolve the data request
+    pub commits: HashSet<OutputPointer>,
+    /// List of reveals to the commitments (contains the data request witnet result)
+    pub reveals: HashSet<OutputPointer>,
+    /// Tally of data request (contains final result)
+    pub tally: Option<OutputPointer>,
+}
+
+/// State of data requests in progress (stored in memory)
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct DataRequestState {
+    /// Data request output (contains all required information to process it)
+    pub data_request: DataRequestOutput,
+    /// List of outputs related to this data request
+    pub info: DataRequestInfo,
+    /// Current stage of this data request
+    pub stage: DataRequestStage,
+    /// The epoch on which this data request has been or will be unlocked
+    // (necessary for removing from the data_requests_by_epoch map)
+    pub epoch: Epoch,
+}
+
+impl DataRequestState {
+    /// Add a new data request state
+    pub fn new(data_request: DataRequestOutput, epoch: Epoch) -> Self {
+        let info = DataRequestInfo::default();
+        let stage = DataRequestStage::COMMIT;
+
+        Self {
+            data_request,
+            info,
+            stage,
+            epoch,
+        }
+    }
+
+    /// Add commit
+    pub fn add_commit(&mut self, output_pointer: OutputPointer) {
+        assert_eq!(self.stage, DataRequestStage::COMMIT);
+        self.info.commits.insert(output_pointer);
+    }
+
+    /// Add reveal
+    pub fn add_reveal(&mut self, output_pointer: OutputPointer) {
+        assert_eq!(self.stage, DataRequestStage::REVEAL);
+        self.info.reveals.insert(output_pointer);
+    }
+
+    /// Add tally and return the data request report
+    pub fn add_tally(
+        mut self,
+        output_pointer: OutputPointer,
+    ) -> (DataRequestOutput, DataRequestReport) {
+        assert_eq!(self.stage, DataRequestStage::TALLY);
+        self.info.tally = Some(output_pointer);
+        // This try_from can only fail if the tally is None, and we have just set it to Some
+        (
+            self.data_request,
+            DataRequestReport::try_from(self.info).unwrap(),
+        )
+    }
+
+    /// Advance to the next stage, returning true on success.
+    /// Since the data requests are updated by looking at the transactions from a valid block,
+    /// the only issue would be that there were no commits in that block.
+    pub fn update_stage(&mut self) -> bool {
+        let old_stage = self.stage;
+
+        self.stage = match self.stage {
+            DataRequestStage::COMMIT => {
+                if self.info.commits.is_empty() {
```

Comment:
In the future, when we verify that there have been enough commitments, should we make the check here? If so, add `TODO`.
```suggestion
                if self.info.commits.is_empty() {
```

Reply:
I added it in `validations.rs`, as it should be at block level. If there are not enough commits for a given data request, then the block should not be considered as valid.